Refactoring of the marine B-matrix job

guillaumevernieres opened the pull request:
# Description
Refactoring of the functionality of the B-matrix generation from the GDASApp.
 
  Resolves #2743 
  
**We need to wait for the companion PR** [GDASApp/pull/1211](https://github.com/NOAA-EMC/GDASApp/pull/1211) **to be merged and the gdas.cd hash to be updated before testing**

What's left to do:
- [x] the jjob dependency has not been fully addressed yet
- [x] remove the old marine B-matrix jjob
- [x] post (in the gdasapp) needs to be updated
- [x] the archiving   "                "
- [x] fix files are resolution dependent 
- [x] use the fix files from glopara in the ci test
- [x] test the cycling at 5.00 deg resolution    **Tested with the same g-w ci setup, except for fit2obs and arch**
- [x] test the cycling 0.25 deg resolution
- [x] test the gfs cycle
- [x] update the `gdas.cd` #

# Type of change
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
For now just the jjob ctests from within the GDASApp

# Checklist
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

guillaumevernieres committed:
3dfgat gdas.cd ctests work but needs tidy

guillaumevernieres committed:
Merge branch 'develop' into feature/marine-det_gv

guillaumevernieres committed:
pynorms

guillaumevernieres committed:
point to different soca tools

guillaumevernieres committed:
added perturbations for envar

guillaumevernieres committed:
working hybvar

CoryMartin-NOAA commented on jobs/JGLOBAL_MARINE_BMAT:
```diff
@@ -0,0 +1,68 @@
+source "${HOMEgfs}/ush/preamble.sh"
+
+export DATA="${DATAROOT}/${RUN}marinebmat_${cyc}"
+export DATA=${DATA:-${DATAROOT}/${RUN}marinebmat_${cyc}}
```
Guessing the first one needs to be removed here?

guillaumevernieres replied:
good catch

CoryMartin-NOAA commented on jobs/JGLOBAL_MARINE_BMAT:
```diff
@@ -0,0 +1,68 @@
+   COMIN_OCEAN_HISTORY_PREV:COM_OCEAN_HISTORY_TMPL \
+   COMIN_ICE_HISTORY_PREV:COM_ICE_HISTORY_TMPL
+
+RUN="enkfgdas" YMD=${gPDY} HH=${gcyc} declare_from_tmpl -rx \
```
`GDUMP_ENS`?

guillaumevernieres replied:
ok

CoryMartin-NOAA commented on sorc/link_workflow.sh:
```diff
@@ -78,7 +78,7 @@ case "${machine}" in
   *)
     echo "FATAL: Unknown target machine ${machine}, couldn't set FIX_DIR"
-    exit 1
+    #exit 0
```
Reminder to revert this

guillaumevernieres replied:
I kept telling myself to not commit ... but here we are ... good catch again.

CoryMartin-NOAA commented on ush/python/pygfs/task/marine_bmat.py:
```diff
@@ -0,0 +1,364 @@
+        logger.info(f"Staging SOCA fix files from {self.task_config.SOCA_INPUT_FIX_DIR}")
+        newdirs = [os.path.join(self.task_config.DATA, 'INPUT')]
+        FileHandler({'mkdir': newdirs}).sync()
+        soca_fix_list = parse_j2yaml(self.task_config.SOCA_FIX_YAML_TMPL, self.task_config)
```
add the `mkdir: [ '{{ DATA }}/INPUT' ]` to `SOCA_FIX_YAML_TMPL`?

guillaumevernieres replied:
yes

CoryMartin-NOAA commented on ush/python/pygfs/task/marine_bmat.py:
```diff
@@ -0,0 +1,364 @@
+                         os.path.join(self.task_config.DATA, "INPUT", "MOM.res.nc")])
+        rst_list.append([os.path.join(self.task_config.DATA, "ice.bkg.nc"),
+                         os.path.join(self.task_config.DATA, "INPUT", "cice.res.nc")])
+        FileHandler({'copy': rst_list}).sync()
```
Combine this and above into a new  Jinja template for FileHandler, Related, not quite clear to me (maybe need a comment), why copy A to B and B to C and not just A to C.

guillaumevernieres replied:
Yes ... it's a mess. I cleaned up the staging of the backgrounds.

CoryMartin-NOAA commented on ush/python/pygfs/task/marine_bmat.py:
```diff
@@ -0,0 +1,364 @@
+
+        # generate the variance partitioning YAML file
+        logger.debug("Generate variance partitioning YAML file")
+        diagb_config = parse_j2yaml(path=os.path.join(self.task_config.BERROR_YAML_DIR, 'soca_diagb.yaml.j2'),
```
I think instead of defining BERROR_YAML_DIR and then hard-coded filenames, these should be user definable through the config.${job} file

guillaumevernieres replied:
You're probably right, but for this PR I'd like to keep things hardcoded.

CoryMartin-NOAA commented on ush/python/pygfs/task/marine_bmat.py:
```diff
@@ -0,0 +1,364 @@
+            diffhz_config.save(os.path.join(self.task_config.DATA, 'soca_parameters_diffusion_hz.yaml'))
+
+        # hybrid EnVAR case
+        if self.task_config.DOHYBVAR == "YES" or self.task_config.NMEM_ENS > 3:
```
Members greater than 3, why can't three members work?

guillaumevernieres replied:
it should, this is dating back to what we were doing with bump. Changed so that 3 should work.

CoryMartin-NOAA commented on ush/python/pygfs/utils/marine_da_utils.py:
```diff
@@ -0,0 +1,100 @@
+logger = getLogger(__name__.split('.')[-1])
+
+@logit(logger)
+def run(exec_cmd):
```
Not immediately clear to me why this is needed, wxflow has an Execute (or Executable or something like that)

guillaumevernieres replied:
exec_cmd is an instance of Executable, run is just a simple wrapper for error trapping and logging. I'll have a look at what's available in Executable.

guillaumevernieres committed:
addressed some of the review comments

guillaumevernieres committed:
Merge branch 'develop' into feature/marine-det_gv

guillaumevernieres committed:
minor tidy

guillaumevernieres committed:
fixed shell norms?

guillaumevernieres committed:
more shell norms

AndrewEichmann-NOAA commented on jobs/JGLOBAL_MARINE_BMAT:
```diff
@@ -0,0 +1,68 @@
+export gPDY=${GDATE:0:8}
+export gcyc=${GDATE:8:2}
+export GDUMP="gdas"
+export GDUMP_ENS="enkf${GDUMP}"
```
The need to export these will be interrogated - `gPDY` and `gcyc` in the python scripts derived from `self.task_config.previous_cycle`

AndrewEichmann-NOAA commented on jobs/rocoto/marinebmat.sh:
```diff
@@ -0,0 +1,25 @@
+###############################################################
+# setup python path for workflow utilities and tasks
+wxflowPATH="${HOMEgfs}/ush/python:${HOMEgfs}/ush/python/wxflow/src"
+socaToolsPATH="${HOMEgfs}/sorc/gdas.cd/ush/soca/tools"
```
Puzzled why these two are needed

guillaumevernieres committed:
job deps and arch

guillaumevernieres committed:
removed dpdcy to soca calc_scales module

guillaumevernieres committed:
removed dpdcy to soca calc_scales module 2

guillaumevernieres committed:
Merge branch 'develop' into feature/marine-det_gv

guillaumevernieres committed:
multiple res

guillaumevernieres committed:
Merge branch 'develop' into feature/marine-det_gv

guillaumevernieres committed:
revert account

guillaumevernieres committed:
Merge branch 'develop' into feature/marine-det_gv

CoryMartin-NOAA commented on ci/cases/gfsv17/ocnanal.yaml:
```diff
@@ -14,9 +14,10 @@ base:
 
 ocnanal:
-  SOCA_INPUT_FIX_DIR: /scratch2/NCEPDEV/ocean/Guillaume.Vernieres/data/static/1440x1080x75/soca
+  SOCA_INPUT_FIX_DIR: /scratch1/NCEPDEV/global/glopara/fix/gdas/soca/20240624/1440x1080x75/soca
```
There should be a machine agnostic way to do this, probably something like `{{ FIXgfs }}/gdas/soca/1440x1080x75/soca`

guillaumevernieres replied:
Ha! Thanks for the pointer!

CoryMartin-NOAA replied:
This may require a change/addition to `sorc/link_workflow.sh` to make it happen though

guillaumevernieres replied:
using your FIXgfs suggestion, thank @CoryMartin-NOAA  ... The cycle isn't complete yet, but it worked for the bmat task ... I hope I didn't jinx it by committing too early!

CoryMartin-NOAA commented on parm/config/gfs/yaml/defaults.yaml:
```diff
@@ -45,13 +45,9 @@ snowanl:
 ocnanal:
-  SOCA_INPUT_FIX_DIR: "/scratch2/NCEPDEV/ocean/Guillaume.Vernieres/data/static/72x35x25/soca"  # TODO: These need to go to glopara fix space.
-  CASE_ANL: "C48"  # TODO: Check in gdasapp if used anywhere for SOCA
+  SOCA_INPUT_FIX_DIR: "/scratch1/NCEPDEV/global/glopara/fix/gdas/soca/20240624/72x35x25/soca"
```
same comment as above

CoryMartin-NOAA commented on ush/python/pygfs/task/marine_bmat.py:
```diff
@@ -0,0 +1,351 @@
+        # Copy diag B files to ROTDIR
+        logger.info(f"Copying diag B files to the ROTDIR")
+        diagb_list = []
+        window_end_iso = self.task_config.MARINE_WINDOW_END.strftime('%Y-%m-%dT%H:%M:%SZ')
```
any reason why the time is the end of the window? wouldn't that be confusing since the model will restart at the beginning of the window?

guillaumevernieres replied:
Yes, I'm using the forecast(s) the furthest away from the IC's for these reasons:
1 - get a few more hours of model integration to get states that are better "balanced"
2 - I expect the ocean spread to be tiny, we're only doing 3D ensda, so I'm picking the ensemble at the end of the window hopping to have slightly more spread.

guillaumevernieres committed:
host dpdt fix dir

RussTreadon-NOAA commented on ush/python/pygfs/utils/__init__.py:
Why is empty file `ush/python/pygfs/utils/__init__.py` added by this PR?

CoryMartin-NOAA replied:
That's required to `import` from a subdirectory. It's a python thing.

RussTreadon-NOAA replied:
Thank you @CoryMartin-NOAA

RussTreadon-NOAA commented on workflow/rocoto/tasks.py:
```diff
@@ -15,7 +15,7 @@ class Tasks:
                    'prepatmiodaobs', 'atmanlinit', 'atmanlvar', 'atmanlfv3inc', 'atmanlfinal',
                    'prepoceanobs',
-                   'ocnanalprep', 'ocnanalbmat', 'ocnanalrun', 'ocnanalecen', 'ocnanalchkpt', 'ocnanalpost', 'ocnanalvrfy',
+                   'ocnanalprep', 'marinebmat', 'ocnanalrun', 'ocnanalecen', 'ocnanalchkpt', 'ocnanalpost', 'ocnanalvrfy',
```
Just curious, all the other Marine DA jobs begin with `ocn`.  The bmat job begins with `marine`  Will the remaining `ocn` jobs eventually be renamed `marine`?

RussTreadon-NOAA reviewed (commented):
Changes look reasonable.  I did not install this PR and ensure jobs run as intended.  Only have two minor questions
1. why add empty file `ush/python/pygfs/utils/__init__.py`?
2. refactored marine B-matrix job uses `marine` as the job prefix.  Will other `ocn` jobs also be renamed in future PR(s)?

guillaumevernieres commented:
> Changes look reasonable. I did not install this PR and ensure jobs run as intended. Only have two minor questions
> 
> 1. why add empty file `ush/python/pygfs/utils/__init__.py`?
> 2. refactored marine B-matrix job uses `marine` as the job prefix.  Will other `ocn` jobs also be renamed in future PR(s)?

Thanks for hte review @RussTreadon-NOAA .
1. marks the directory as a package
2. yes, we're in the process of switching to using "marine"

RussTreadon-NOAA reviewed (dismissed):
Approve.

aerorahul commented on jobs/JGLOBAL_MARINE_BMAT:
```diff
@@ -0,0 +1,67 @@
+
+source "${HOMEgfs}/ush/preamble.sh"
+
+export DATA=${DATA:-${DATAROOT}/${RUN}marinebmat_${cyc}}
+# source config.base, config.ocnanal and config.marinebmat
+# and pass marinebmat to ${machine}.env
+source "${HOMEgfs}/ush/jjob_header.sh" -e "marinebmat" -c "base ocnanal marinebmat"
+
+# If ensemble perturbations are used, save them in a DATAenspert directory that will
+# exist throughout the DA cycle
+if (( 10#${NMEM_ENS:-0} > 0 )); then
+  export DATAjob="${DATAROOT}/${RUN}marinebmat.${PDY:-}${cyc}"
+  export DATAenspert="${DATAjob}/enspert"
+  if [[ ! -d "${DATAenspert}" ]]; then mkdir -p "${DATAenspert}"; fi
+fi
```
```suggestion
if (( 10#${ENSMEM:-0} > 0 )); then
  export DATAjob="${DATAROOT}/${RUN}marinebmat.${PDY:-}${cyc}"
  export DATA="${DATAjob}/${jobid}"
  # Create the directory to hold ensemble perturbations
  export DATAenspert="${DATAjob}/enspert"
  if [[ ! -d "${DATAenspert}" ]]; then mkdir -p "${DATAenspert}"; fi
fi

# source config.base, config.ocnanal and config.marinebmat
# and pass marinebmat to ${machine}.env
source "${HOMEgfs}/ush/jjob_header.sh" -e "marinebmat" -c "base ocnanal marinebmat"
```

aerorahul commented on ci/cases/gfsv17/ocnanal.yaml:
```diff
@@ -14,9 +14,10 @@ base:
 
 ocnanal:
-  SOCA_INPUT_FIX_DIR: /scratch2/NCEPDEV/ocean/Guillaume.Vernieres/data/static/1440x1080x75/soca
+  SOCA_INPUT_FIX_DIR: {{ FIXgfs }}/gdas/soca/1440x1080x75/soca
```
Does this work?  `FIXgfs` is not available at the setup level.  It is available in the jobs.  This should be moved to the config files.

guillaumevernieres replied:
Just tested again, it does work.  `{{ FIXgfs }}/gdas/soca/1440x1080x75/soca` becomes `${ FIXgfs}/gdas/soca/1440x1080x75/soca` in the config under exp dir. 

I want to keep the option exposed in the yaml file, since that fix dir is still fluctuating ...

aerorahul commented on jobs/JGLOBAL_ARCHIVE:
```diff
@@ -35,6 +35,8 @@ YMD=${PDY} HH=${cyc} declare_from_tmpl -rx \
 	COMIN_OCEAN_GRIB:COM_OCEAN_GRIB_TMPL \
 	COMIN_OCEAN_NETCDF:COM_OCEAN_NETCDF_TMPL \
 	COMIN_OCEAN_ANALYSIS:COM_OCEAN_ANALYSIS_TMPL \
+        COMIN_OCEAN_BMATRIX:COM_OCEAN_BMATRIX_TMPL \
+        COMIN_ICE_BMATRIX:COM_ICE_BMATRIX_TMPL \
```
```suggestion
    COMIN_OCEAN_BMATRIX:COM_OCEAN_BMATRIX_TMPL \
    COMIN_ICE_BMATRIX:COM_ICE_BMATRIX_TMPL \
```

aerorahul commented on parm/archive/gdasocean_analysis.yaml.j2:
```diff
@@ -3,23 +3,30 @@ gdasocean_analysis:
-        {% if NMEM_ENS > 2 %}
-        - '{{ COMIN_OCEAN_ANALYSIS | relpath(ROTDIR) }}/{{ head }}{{domain}}.recentering_error.nc'
-        {% endif %}
+            - '{{ COMIN_OCEAN_ANALYSIS | relpath(ROTDIR) }}/{{ head }}{{domain}}.incr.nc'
+            - '{{ COMIN_OCEAN_ANALYSIS | relpath(ROTDIR) }}/{{ head }}{{domain}}ana.nc'
```
Is the `ice.incr.nc` and `ice.ana.nc` in the `COM_OCEAN_ANALYSIS`?  I would think they should be in `COM_ICE_ANALYSIS`.

guillaumevernieres replied:
Will do this when refactoring the deterministic analysis.

aerorahul commented on ci/cases/gfsv17/ocnanal.yaml:
```diff
@@ -25,6 +26,6 @@ ocnanal:
 
 prepoceanobs:
-  SOCA_OBS_LIST: {{ HOMEgfs }}/sorc/gdas.cd/parm/soca/obs/obs_list.yaml 
+  SOCA_OBS_LIST: {{ HOMEgfs }}/sorc/gdas.cd/parm/soca/obs/obs_list.yaml
```
Should be moved to `config.prepoceanobs`. 
Same in `config.ocnanl`.
Also see comment below.

guillaumevernieres replied:
What's wrong with keeping this exposed in the yaml?

aerorahul commented on jobs/JGLOBAL_MARINE_BMAT:
```diff
@@ -0,0 +1,67 @@
+###############################################################
+# Run relevant script
+
+EXSCRIPT=${GDASMARINEBMATRUNPY:-${SCRgfs}/exglobal_marine_bmat_run.py}
```
why `_run`?
EE2 will make us change this to match the j-job name.

guillaumevernieres replied:
ok, will push the changes after I'm done testing

aerorahul commented on parm/config/gfs/config.ocnanal:
```diff
@@ -6,21 +6,15 @@
-export OBS_LIST=@SOCA_OBS_LIST@
-export OBS_YAML="${OBS_LIST}"
-export FV3JEDI_STAGE_YAML="${HOMEgfs}/sorc/gdas.cd/test/soca/testinput/dumy.yaml"
+export OBS_LIST=@SOCA_OBS_LIST@  # TODO(GA): doesn't look necessary as is to have
```
Based on the comments able, can this become:
```suggestion
export OBS_LIST="${PARMgfs}/gdas/soca/obs/obs_list.yaml"
```
Same change is likely needed in `config.prepoceanobs`

guillaumevernieres replied:
same comment as above, I want to have the option to point to a different obs list in that yaml.

aerorahul commented on parm/config/gfs/config.ocnanal:
```diff
@@ -6,21 +6,15 @@
-export FV3JEDI_STAGE_YAML="${HOMEgfs}/sorc/gdas.cd/test/soca/testinput/dumy.yaml"
+export OBS_LIST=@SOCA_OBS_LIST@  # TODO(GA): doesn't look necessary as is to have
+export OBS_YAML="${OBS_LIST}"    #           OBS_LIST and OBS_YAML pick one or add logic
 export SOCA_INPUT_FIX_DIR=@SOCA_INPUT_FIX_DIR@
```
This need not be this way. 
See above comment.
If there is a dependence on the number of levels, we can make a case statement here and cover all cases.

aerorahul commented on parm/config/gfs/yaml/defaults.yaml:
```diff
@@ -45,16 +45,12 @@ snowanl:
 
 prepoceanobs:
-  SOCA_INPUT_FIX_DIR: "/scratch2/NCEPDEV/ocean/Guillaume.Vernieres/data/static/72x35x25/soca"  # TODO: These need to go to glopara fix space.
+  SOCA_INPUT_FIX_DIR: "${FIXgfs}/gdas/soca/72x35x25/soca"
```
This should be moved to `config.prepoceanobs` since it is now in glopara space

aerorahul commented on parm/config/gfs/yaml/defaults.yaml:
```diff
@@ -45,16 +45,12 @@ snowanl:
 ocnanal:
-  SOCA_INPUT_FIX_DIR: "/scratch2/NCEPDEV/ocean/Guillaume.Vernieres/data/static/72x35x25/soca"  # TODO: These need to go to glopara fix space.
-  CASE_ANL: "C48"  # TODO: Check in gdasapp if used anywhere for SOCA
+  SOCA_INPUT_FIX_DIR: "${FIXgfs}/gdas/soca/72x35x25/soca"
```
Should be moved to `config.ocnanal` since this is now in glopara space.

aerorahul commented on parm/config/gfs/yaml/defaults.yaml:
```diff
@@ -45,16 +45,12 @@ snowanl:
 prepoceanobs:
-  SOCA_INPUT_FIX_DIR: "/scratch2/NCEPDEV/ocean/Guillaume.Vernieres/data/static/72x35x25/soca"  # TODO: These need to go to glopara fix space.
+  SOCA_INPUT_FIX_DIR: "${FIXgfs}/gdas/soca/72x35x25/soca"
   SOCA_OBS_LIST: "${PARMgfs}/gdas/soca/obs/obs_list.yaml"  # TODO: This is also repeated in ocnanal
   OBSPREP_YAML: "${PARMgfs}/gdas/soca/obsprep/obsprep_config.yaml"
```
These need not be in defaults, as they are in HOMEgfs.
Consider moving them to `config.ocnanal` and `config.prepobs` as needed.

aerorahul commented on ush/python/pygfs/utils/marine_da_utils.py:
```diff
@@ -0,0 +1,105 @@
+        raise OSError(f"Failed to execute {exec_cmd}")
+    except Exception:
+        raise WorkflowException(f"An error occured during execution of {exec_cmd}")
+
+    pass
```
```suggestion
```
No need for `pass`.

aerorahul commented on ush/python/pygfs/utils/marine_da_utils.py:
```diff
@@ -0,0 +1,105 @@
+    """Link the executable to the DATA directory
+    """
+    logger.info(f"Link executable {exe_name}")
+    logger.warn("Linking is not permitted per EE2.")
```
```suggestion
    logger.warn("WARNING: Linking is not permitted per EE2.")
```
EE2 suggested way to warn.

aerorahul commented on ush/python/pygfs/utils/marine_da_utils.py:
```diff
@@ -0,0 +1,105 @@
+@logit(logger)
+def prep_input_nml(task_config) -> None:
+    """Prepare the input.nml file
+       TODO: Use jinja2 instead of f90nml
```
👍🏾

aerorahul commented on ush/python/pygfs/utils/marine_da_utils.py:
```diff
@@ -0,0 +1,105 @@
+
+
+@logit(logger)
+def run(exec_cmd):
```
We will try to elevate this to `wxflow.Executable` for a short-hand to use by all.  
This is repeated in many places.

guillaumevernieres replied:
ok.

aerorahul commented on ush/python/pygfs/utils/marine_da_utils.py:
```diff
@@ -0,0 +1,105 @@
+import f90nml
```
Please use `type hints` to provide input/output types in the functions/classes.
While dynamic typing offers flexibility, it also creates room for potential bugs.
type hinting can significantly enhance code readability, help with documenting and debugging to ensure the right datatypes are being passed.

guillaumevernieres replied:
Ok. Will push the changes after testing.

aerorahul reviewed (changes requested):
some comments from review.

guillaumevernieres committed:
Update jobs/JGLOBAL_MARINE_BMAT

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
Update jobs/JGLOBAL_ARCHIVE

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
cycle on orion/hercules

guillaumevernieres committed:
Merge branch 'feature/marine-det_gv' of https://github.com/guillaumevernieres/global-workflow into feature/marine-det_gv

guillaumevernieres committed:
Update ush/python/pygfs/utils/marine_da_utils.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
renamed ex script

guillaumevernieres committed:
Merge branch 'feature/marine-det_gv' of https://github.com/guillaumevernieres/global-workflow into feature/marine-det_gv

guillaumevernieres committed:
Update ush/python/pygfs/utils/marine_da_utils.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
...

guillaumevernieres committed:
Merge branch 'feature/marine-det_gv' of https://github.com/guillaumevernieres/global-workflow into feature/marine-det_gv

guillaumevernieres committed:
Merge branch 'NOAA-EMC:feature/marine-det_gv' into feature/marine-det_gv

guillaumevernieres committed:
fixed missing import

guillaumevernieres committed:
Merge branch 'feature/marine-det_gv' of https://github.com/guillaumevernieres/global-workflow into feature/marine-det_gv

guillaumevernieres commented:
I addressed some of your comments @aerorahul . The one I did not (obs related) will be addressed when refactoring the deterministic analysis, which should happen shortly after this PR is merged.

guillaumevernieres committed:
Merge branch 'develop' into feature/marine-det_gv

guillaumevernieres committed:
Merge branch 'develop' into feature/marine-det_gv

guillaumevernieres commented:
@aerorahul Can you give me a heads up before triggering the ci on this PR? I need to merge https://github.com/NOAA-EMC/GDASApp/pull/1211 and update the gdas.cd #, but I'd like to do this at the last minute, since this is going to break gdas.cd develop.

guillaumevernieres committed:
new hash for gdas.cd

guillaumevernieres commented:
FYI @aerorahul , I just updated the gdas.cd commit.

DavidHuber-NOAA commented:
Starting CI.

CoryMartin-NOAA commented:
I'm not quite sure exactly what is going on with the snow DA failure. I suspect it is due to an updated GDAS hash but @jiaruidong2017 's workflow PR is not yet merged in. I think his PR will need to go in first to make these tests pass.

jiaruidong2017 commented:
@CoryMartin-NOAA is right. Because the updates from the GDASApp's PR is merged into develop, but the updates from the G-W's PR is not merged in yet.

guillaumevernieres commented:
> Experiment C48mx500_3DVarAOWCDA **FAILED** on Hera in `/scratch1/NCEPDEV/global/CI/2749/RUNTESTS/EXPDIR/C48mx500_3DVarAOWCDA_4ba7dee1`

The archiving task failed ... Oops.

DavidHuber-NOAA commented:
@guillaumevernieres It looks like the `gdasprepsnowobs` job also failed to locate an input file.

guillaumevernieres commented:
> @guillaumevernieres It looks like the `gdasprepsnowobs` job also failed to locate an input file.

I'll coordinate with @jiaruidong2017 and let you know once we have fixes for the marine and snow issues @DavidHuber-NOAA .

jiaruidong2017 commented:
> /scratch1/NCEPDEV/global/CI/2749/RUNTESTS/COMROOT/C96_atmaerosnowDA_4ba7dee1/logs/2021122018/gdasprepsnowobs.log



> > @guillaumevernieres It looks like the `gdasprepsnowobs` job also failed to locate an input file.
> 
> I'll coordinate with @jiaruidong2017 and let you know once we have fixes for the marine and snow issues @DavidHuber-NOAA .

The CI test depends on https://github.com/NOAA-EMC/global-workflow/pull/2779. We moved the ims pre-processing from 18z to 00z.

guillaumevernieres committed:
snow at 18z and fixed marine da arch j2

guillaumevernieres committed:
Merge branch 'develop' into feature/marine-det_gv

guillaumevernieres commented:
@jiaruidong2017 , can you review this PR and check the changes to the snow script? I did not merge your branch into mine, I just did a checkout of your file from your branch, so git blame will still point to you!

@DavidHuber-NOAA , I was not able to test the archive task but I did find indentation bugs in the jinja template. Can you select what ci test to run? If so, any chance you could only trigger the ones that failed for now?

jiaruidong2017 reviewed (approved):
I confirmed all my changes from the PR #2779 were included here.

DavidHuber-NOAA commented:
@guillaumevernieres Unfortunately, I cannot control which tests run when using the CI system.  I will go ahead and start running tests on WCOSS2 and Hercules.  Hera's CI system is currently offline, so I will run those manually.

EDIT: On hera, I will just run the SnowDA and AOWCDA tests.

DavidHuber-NOAA commented:
The errors reported in CI are mostly my fault, but there is one change needed to `config/archive/gdas.yaml.j2` at line 25.

```yaml
        - "logs/{{ cycle_YMDH }}/{{ RUN }}ocnanalbmat.log"
```

Should be changed to

```yaml
        - "logs/{{ cycle_YMDH }}/{{ RUN }}marinebmat.log"
```

DavidHuber-NOAA commented:
I have made this change locally on Hera, and the tests ran to completion.  I will submit a PR to your branch @guillaumevernieres.  

WCOSS2 tests needed to be tweaked this morning and are running again.

I'm not sure why tests did not run on Hercules.  I will investigate that further.  That said, I think that if Hera and WCOSS2 pass, then this PR should be merged.

guillaumevernieres commented:
> I have made this change locally on Hera, and the tests ran to completion. I will submit a PR to your branch @guillaumevernieres.
> 
> WCOSS2 tests needed to be tweaked this morning and are running again.
> 
> I'm not sure why tests did not run on Hercules. I will investigate that further. That said, I think that if Hera and WCOSS2 pass, then this PR should be merged.

Thanks for looking into this @DavidHuber-NOAA . Did you run the tests on hera already or is that still pending?

DavidHuber-NOAA committed:
Rename marinebmat log file in gdas.yaml.j2

DavidHuber-NOAA commented:
The tests on Hera are complete and successful.  Updating the label and submitting a PR now.

DavidHuber-NOAA commented:
@malloryprow The `gfsmetpg2o` job is failing for the `C48_ATM` and `C48_S2SW` tests on WCOSS2 because the NAM `prepbufr` data is not available in `/lfs/h2/emc/vpppg/noscrub/emc.vpppg/verification/global/archive/obs_data/prepbufr/nam` for cycles `2021032312` through `2021032400`.  Would you be able to upload this data for these dates to WCOSS2?

DavidHuber-NOAA commented:
@guillaumevernieres PR https://github.com/guillaumevernieres/global-workflow/pull/7 opened.

malloryprow commented:
Pulling the data from HPSS now. I'll let you know when it is in place.

> @malloryprow The `gfsmetpg2o` job is failing for the `C48_ATM` and `C48_S2SW` tests on WCOSS2 because the NAM `prepbufr` data is not available in `/lfs/h2/emc/vpppg/noscrub/emc.vpppg/verification/global/archive/obs_data/prepbufr/nam` for cycles `2021032312` through `2021032400`. Would you be able to upload this data for these dates to WCOSS2?

guillaumevernieres committed:
Merge pull request #7 from DavidHuber-NOAA/fix_gdas_tmpl

Rename marinebmat log file in gdas.yaml.j2

DavidHuber-NOAA commented:
@malloryprow I believe we also need the GDAS `prepbufr` files for the same date range.  Similarly, we need GDAS and NAM files for cycles `2021122100` through `202112200` for the `C96C48_hybatmDA` test.

malloryprow commented:
@DavidHuber-NOAA The requested prepbufr files should be in place now.

DavidHuber-NOAA commented:
Great, thank you @malloryprow!

DavidHuber-NOAA commented:
All tests ran to completion successfully with the exception of the gfsmetpg2g1 and gfsmetpg2o1 jobs for the first two GFS cycles of the `C96_atm3DVar_extended` and `C96C48_hybatm` tests.  I think these can be ignored for now and I will address these in another issue/PR.
